Source/DafnyLanguageServer/Language/Symbols/SymbolTableFactory.cs

```diff
@@ -179,11 +219,17 @@ private class DesignatorVisitor : SyntaxTreeVisitor {
           // Many resolutions for automatically generated nodes (e.g. Decreases, Update when initializating a variable
           // at declaration) cause duplicated visits. These cannot be prevented at this time as it seems there's no way
           // to distinguish nodes from automatically created one (i.e. nodes of the original syntax tree vs. nodes of the
-          // abstract syntax tree). We could just ignore such duplicates. However, we may miss programatic errors if we
-          // do so.
+          // abstract syntax tree). We just ignore such duplicates until more information is availabe in the AST.
           var range = token.GetLspRange();
           SymbolLookup.Add(range.Start, range.End, symbol);
-          designators.Add(node, symbol);
+          if (designators.ContainsKey(node)) {
+            if (designators[node] != symbol) {
```

Comment:
I would definitely prefer to use TryGetValue, however, I did this change so that I could be able to determine when this case happens. Here is one simple case that is impossible to detect and that happens:

```
datatype Pos = Pos(line: int)
function method f(i: int): Pos {
  if i <= 3 then Pos(i)
  else
   var r := f(i - 2);
   Pos(r.line + 2)
}
```

It is impossible to distinguish between this code above and replacing it with `var r: Pos := f(i - 2)`, because there is no information that the type was actually inferred.
But in the first case, it will still put a type there and will reuse the same type token as the function's output type, which led to crashes in my editor whenever I was relying on type inference.
One day, the AST will be better aware of which elements are user-generated and inferred, but for now, it's better to be fail-safe as you say.
I added the check that the added element should have the same symbol, which should help us find cases which are really problematic.

Reply:
Wouldn't this work?
```cs
          SymbolLookup.Add(range.Start, range.End, symbol);
          if (designators.TryGetValue(node, out var registeredSymbol)) {
            if (registeredSymbol != symbol) {
              logger.LogInformation("Conflicting  symbol resolution nf designator named {Identifier} in {Filename}@({Line},{Column})",
                identifier, token.GetDocumentFileName(), token.line, token.col);
            }
          } else {
            designators.Add(node, symbol);
          }
```

I see now the necessity for `FirstOrDefault` in `SymbolTable`. However, I think it's better we prevent the duplicate registration here than handling this case in the `SymbolTable`.

Reply:
I am now using TryGetValue instead of querying the dictionary twice